src/osd/ECCommon.cc

```diff
@@ -564,13 +559,29 @@ struct ClientReadCompleter : ECCommon::ReadCompleter {
         goto out;
       }
       bufferlist trimmed;
-      auto off = read.offset - aligned.first;
-      auto len = std::min(read.size, bl.length() - off);
-      dout(20) << __func__ << " bl.length()=" << bl.length()
-	       << " len=" << len << " read.size=" << read.size
-	       << " off=" << off << " read.offset=" << read.offset
-	       << dendl;
-      trimmed.substr_of(bl, off, len);
+      // If partial stripe reads are disabled read_offset_in_stripe will be 0
+      // which will mean trim_offset is 0. When partial reads are enabled the
+      // shards read (wanted_to_read) is a union of the requirements for each
+      // stripe, each range being read may need to trim unneeded shards
+      uint64_t read_offset_in_stripe = read.offset -
+	read_pipeline.sinfo.logical_to_prev_stripe_offset(read.offset);
+      uint64_t chunk_size = read_pipeline.sinfo.get_chunk_size();
+      uint64_t trim_offset = 0;
+      for (auto shard: wanted_to_read) {
+	if (shard * chunk_size < read_offset_in_stripe) {
```

Comment:
Should this be `<=`?  Otherwise, if `read_offset_in_stripe` is at the shard boundary, it seems like the preceding shard would not be trimmed.

Reply:
The code works and < is correct, but only because `read.offset` is always aligned to a shard boundary and hence `read_offset_in_stripe` is also aligned. This means we are comparing the start offset of each chunk that has been read to the start offset of the chunk we want to keep - if it is < then we need to trim it, if it is = we want to keep it.

The alignment to shard boundaries - either to chunk boundaries for partial reads or to stripe boundaries otherwise occurs in ECBackend::objects_read_async (which also has the completion callback that trims the buffer back to the required size). Overwrites and reconstructs also issue reads, but these are always stripe aligned.

But the `bounds` and `aligned` variables above implies that this code is intended to cope with misaligned ranges, even if this currently never happens. If `read_offset_in_stripe` was not aligned then this code would incorrectly trim a shard that contains data we want. So I think that `read_offset_in_stripe` should be renamed as `aligned_offset_in_stripe` and be calculated from `aligned.first` rather than `read.offset`

Reply:
> If `read_offset_in_stripe` was not aligned then this code would incorrectly trim a shard that contains data we want.

Heh, I actually wrote (almost) that but then didn't like the wording, incorrectly inverted it and ended up saying that the "at the shard boundary" case is wrong instead of the "in the middle of the shard" case.  Mea culpa.